core/src/queue.rs

```diff
@@ -147,49 +147,84 @@ impl Queue {
+            Err(e) => Err((tx, e)),
+            Ok(MustUse(signature_check)) => {
+                // Get `txs_len` before entry to avoid deadlock
+                let txs_len = self.txs.len();
+                let hash = tx.hash();
+                let entry = match self.txs.entry(hash) {
+                    Entry::Occupied(mut old_tx) => {
+                        // MST case
+                        old_tx
+                            .get_mut()
+                            .as_mut_v1()
+                            .signatures
+                            .extend(tx.as_v1().signatures.clone());
+                        return Ok(());
+                    }
+                    Entry::Vacant(entry) => entry,
+                };
+                if txs_len >= self.max_txs {
+                    return Err((tx, Error::Full));
                 }
-                Entry::Vacant(entry) => entry,
+
+                // Reason for such insertion order is to avoid situation
+                // when poped from the `queue` hash does not yet has corresponding (hash, tx) record in `txs`
+                entry.insert(tx);
+                let queue_to_push = if signature_check {
+                    &self.queue
+                } else {
+                    &self.waiting_buffer
+                };
+                queue_to_push.push(hash).map_err(|err_hash| {
+                    let (_, err_tx) = self
+                        .txs
+                        .remove(&err_hash)
+                        .expect("Inserted just before match");
+                    (err_tx, Error::Full)
+                })
+            }
+        }
+    }
+
+    /// Pop single transaction from the waiting buffer.
+    ///
+    /// Records unsigned transaction in `seen`.
+    fn pop_from_waiting_buffer(
+        &self,
+        seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
+        wsv: &WorldStateView,
+    ) -> Option<VersionedAcceptedTransaction> {
+        loop {
+            let hash = self.waiting_buffer.pop()?;
+            let entry = match self.txs.entry(hash) {
+                Entry::Occupied(entry) => entry,
+                // As practice shows this code is not `unreachable!()`.
+                // When transactions are submitted quickly it can be reached.
+                Entry::Vacant(_) => continue,
             };
 
-            // Reason for such insertion order is to avoid situation
-            // when poped from the `queue` hash does not yet has corresponding (hash, tx) record in `txs`
-            entry.insert(tx);
-            self.queue.push(hash).map_err(|err_hash| {
-                let (_, err_tx) = self
-                    .txs
-                    .remove(&err_hash)
-                    .expect("Inserted just before match");
-                (err_tx, Error::Full)
-            })
+            match self.check_tx(entry.get(), wsv) {
+                Err(_) => {
+                    entry.remove_entry();
+                    continue;
```

Comment:
Technically this `continue` is redundant. Given that the other pattern is an early return, just reorder the arms and move this to the bottom.